gradle-jdks/src/main/java/com/palantir/gradle/jdks/SynchronizedDirectoryProperty.java

```diff
@@ -0,0 +1,194 @@
+/*
+ * (c) Copyright 2022 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.gradle.jdks;
+
+import java.io.File;
+import java.util.function.BiFunction;
+import javax.annotation.Nullable;
+import org.gradle.api.Transformer;
+import org.gradle.api.file.Directory;
+import org.gradle.api.file.DirectoryProperty;
+import org.gradle.api.file.FileCollection;
+import org.gradle.api.file.FileTree;
+import org.gradle.api.file.RegularFile;
+import org.gradle.api.provider.Provider;
+
+final class SynchronizedDirectoryProperty implements DirectoryProperty {
```

Comment:
My worry about doing this is that if Gradle ever add a method to this interface, everything explodes. I think we can do something similar with `Proxy`s to avoid that case by making the type at runtime.

Reply:
I tried it and it doesn't work - but for the same reason this code doesn't work. Gradle casting interfaces to their own internal types strikes again:

```
java.lang.ClassCastException: class com.palantir.gradle.jdks.SynchronizedDirectoryProperty cannot be cast to class org.gradle.api.internal.provider.PropertyInternal
```